internal/k0sctl/action/apply.go

```diff
@@ -0,0 +1,152 @@
+package action
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/k0sproject/k0sctl/analytics"
+	"github.com/k0sproject/k0sctl/phase"
+
+	provider_phase "github.com/mirantis/terraform-provider-k0sctl/internal/k0sctl/phase"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type Apply struct {
+	// Manager is the phase manager
+	Manager *phase.Manager
+	// DisableDowngradeCheck skips the downgrade check
+	DisableDowngradeCheck bool
+	// Force allows forced installation in case of certain failures
+	Force bool
+	// NoWait skips waiting for the cluster to be ready
+	NoWait bool
+	// NoDrain skips draining worker nodes
+	NoDrain bool
+	// RestoreFrom is the path to a cluster backup archive to restore the state from
+	RestoreFrom string
+	// KubeconfigOut is a writer to write the kubeconfig to
+	KubeconfigOut io.Writer
+	// KubeconfigAPIAddress is the API address to use in the kubeconfig
+	KubeconfigAPIAddress string
+	// ConfigPath is the path to the configuration file (used for kubeconfig command tip on success)
+	ConfigPath string
+}
+
+func (a Apply) Run() error {
+	start := time.Now()
+
+	phase.NoWait = a.NoWait
+	phase.Force = a.Force
+
+	lockPhase := &phase.Lock{}
+
+	a.Manager.AddPhase(
+		&phase.DefaultK0sVersion{},
+		&phase.Connect{},
+		&phase.DetectOS{},
+		lockPhase,
+		&phase.PrepareHosts{},
+		&phase.GatherFacts{},
+		&phase.ValidateHosts{},
+		&phase.GatherK0sFacts{},
+		&phase.ValidateFacts{SkipDowngradeCheck: a.DisableDowngradeCheck},
+
+		// if UploadBinaries: true
+		&phase.DownloadBinaries{}, // downloads k0s binaries to local cache
+		&phase.UploadK0s{},        // uploads k0s binaries to hosts from cache
+
+		// if UploadBinaries: false
+		&phase.DownloadK0s{}, // downloads k0s binaries directly from hosts
+
+		&phase.UploadFiles{},
+		&phase.InstallBinaries{},
+		&phase.PrepareArm{},
+		&phase.ConfigureK0s{},
+		&phase.Restore{
+			RestoreFrom: a.RestoreFrom,
+		},
+		&phase.RunHooks{Stage: "before", Action: "apply"},
+		&phase.InitializeK0s{},
+		&phase.InstallControllers{},
+		&phase.InstallWorkers{},
+		&phase.UpgradeControllers{},
+		&phase.UpgradeWorkers{NoDrain: a.NoDrain},
+		&phase.ResetWorkers{NoDrain: a.NoDrain},
+		&phase.ResetControllers{NoDrain: a.NoDrain},
+		&provider_phase.ValidateHostsExtended{},
+		&phase.RunHooks{Stage: "after", Action: "apply"},
+	)
+
+	if a.KubeconfigOut != nil {
+		a.Manager.AddPhase(&phase.GetKubeconfig{APIAddress: a.KubeconfigAPIAddress})
+	}
+
+	a.Manager.AddPhase(
+		&phase.Unlock{Cancel: lockPhase.Cancel},
+		&phase.Disconnect{},
+	)
+
+	analytics.Client.Publish("apply-start", map[string]interface{}{})
+
+	var result error
+
+	if result = a.Manager.Run(); result != nil {
+		analytics.Client.Publish("apply-failure", map[string]interface{}{"clusterID": a.Manager.Config.Spec.K0s.Metadata.ClusterID})
```

Comment:
You can create a variable and assign it to `map[string]interface{}{"clusterID": a.Manager.Config.Spec.K0s.Metadata.ClusterID`, then re-use the variable as needed

Reply:
apply.go file is directly copy pasted from the k0sctl source code, the only change done is addling the line 82 for appending the new phase. I would like to keep the diffs minimum with the original apply.go in k0sctl apply.go so that it it will be easy for us to merge the changes we added to the provider, back to k0sctl in future.

Reply:
I don't think that DRY should strictly apply here, as the connection between the cases is not absolute. A proper approach woulud be to better wrap the analytics singleton instead of creating any common interfaces here.
We could consider a future ticket for such cleanup.